File: src/StacksAndQueues/CircularQueue.java
package StacksAndQueues;

public class CircularQueue {
    protected int[] data;
    protected int size = 0;
    protected int front = 0;
    protected int end = 0;
    private static final int DEFAULTSIZE = 5;
    public CircularQueue(){
        this(DEFAULTSIZE);
    }
    public CircularQueue(int size){
        data = new int[size];
    }
    public boolean add(int elem) throws CustomStackException {
        if(isFull()){
            throw new CustomStackException("Queue is Full");
        }
        data[end] = elem;
        end+=1;
        end = end% data.length;
        size+=1;
        return true;
    }
    public int remove() throws CustomStackException{
        if(isEmpty()){
            throw new CustomStackException("Queue is Empty");
        }
        int removed = data[front];
        front+=1;
        front = front% data.length;
        size-=1;
        return removed;
    }
    public int peek() throws CustomStackException{
        if(isEmpty()){
            throw new CustomStackException("Queue is Empty");
        }
        return data[front];
    }
    public void display(){
        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }
//        int i=front;
//        do{
//            System.out.print(data[i]+" ");
//            i+=1;
//            i = i% data.length;
//        }while(i!=end);
        for(int i=front;i!=end;){
            System.out.print(data[i]+" ");
            i+=1;
            i = i%data.length;
        }
        System.out.println("END");
    }
    public boolean isFull(){
        if(size==data.length){
            return true;
        }
        return false;
    }
    public boolean isEmpty(){
        if(size==0){
            return true;
        }
        return false;
    }
}

File: pythonFiles/StacksAndQueues/circular_queue.py
class CircularQueue:
    def __init__(self,size=10):
        self.l = [0]*size
        self.size = 0
        self.f = 0
        self.e = 0
    
    def add(self,element):
        if(self.isFull()):
            raise Exception("Queue is Full")
        print("added")
        self.e  = self.e % len(self.l)
        self.l[self.e] = element
        self.e += 1
        self.size += 1
        return True

    def remove(self):
        if(self.isEmpty()):
            raise Exception("Queue is Empty")
        removed = self.l[self.f]
        self.f += 1
        self.f = self.f % len(self.l)
        self.size -= 1
        return removed

    def front(self):
        if(self.isEmpty()):
            raise Exception("Queue is Empty")
        return self.l[self.f]

    def isFull(self):
        if(self.size==len(self.l)):
            return True
        return False
    
    def isEmpty(self):
        if(self.size==0):
            return True
        return False

    def display(self):
        print("Circular Queue:")
        print(self.f,self.e,self.size)
        for i in range(self.f,self.f + self.size):
            print(self.l[i%len(self.l)])

